py/profiling.c

```diff
@@ -0,0 +1,1046 @@
+    }
+
+    if (*code_state->ip == MP_BC_FOR_ITER) {
+        // assert(false);
+    }
+
+    mp_obj_frame_t *o = frame;
+    mp_obj_code_t *code = o->code;
+    const mp_raw_code_t *rc = code->rc;
+    const mp_bytecode_prelude_t *prelude = &rc->prelude;
+    
+    assert(o->code_state == code_state);
+
+    o->lasti = code_state->ip - prelude->bytecode;
+    o->lineno = prof_bytecode_lineno(rc, o->lasti);
+
+    return MP_OBJ_FROM_PTR(o);
+}
+
+
+//
+// TRACE
+//
+
+mp_obj_t prof_settrace(mp_obj_t callback) {
+    // mp_printf(&mp_plat_print, "settraced!!!!!!\n");
+    if (mp_obj_is_callable(callback)) {
+        prof_trace_cb = callback;
+    } else {
+        prof_trace_cb = NULL;
+    }
+    return mp_const_none;
+}
+
+mp_obj_t prof_callback_invoke(mp_obj_t callback, prof_callback_args_t *args) {
+
+    assert(mp_obj_is_callable(callback));
+    
+    prof_is_executing = true;
+
+    mp_obj_t a[3] = {args->frame, args->event, args->arg};
+    mp_obj_t top = mp_call_function_n_kw(callback, 3, 0, a);
+
+    prof_is_executing = false;
+
+    if (MP_STATE_VM(mp_pending_exception) != MP_OBJ_NULL) {
+        mp_obj_t obj = MP_STATE_VM(mp_pending_exception);
+        MP_STATE_VM(mp_pending_exception) = MP_OBJ_NULL;
+        nlr_raise(obj);
+    }
+    return top;
+}
+
+mp_obj_t prof_instr_tick(mp_code_state_t *code_state, bool isException) {
+
+    // Detecet execution recursion.
+    assert(!prof_is_executing);
+    assert(code_state->frame);
+    assert(mp_obj_get_type(code_state->frame) == &mp_type_frame);
+
+    // Detecet data recursion.
+    assert(code_state != code_state->prev_state);
+
+    mp_obj_t top = mp_const_none;
+    mp_obj_t callback = code_state->frame->callback;
+
+    prof_callback_args_t _args, *args=&_args;
+    args->frame = code_state->frame;
+    args->event = mp_const_none;
+    args->arg = mp_const_none;
+
+    // CALL
+    // Call event's are handled inside the vm.c
+
+    // EXCEPTION
+    if (isException) {
+        args->event = MP_ROM_QSTR(MP_QSTR_exception);
+        top = prof_callback_invoke(callback, args);
+        return top;
+    }
+
+    // LINE
+    const mp_raw_code_t *rc = code_state->fun_bc->rc;
+    const mp_bytecode_prelude_t *prelude = &rc->prelude;
+    size_t prev_line_no = args->frame->lineno;
+    size_t current_line_no = prof_bytecode_lineno(rc, code_state->ip - prelude->bytecode);
+    if (prev_line_no != current_line_no) {
+        args->frame->lineno = current_line_no;
+        args->event = MP_ROM_QSTR(MP_QSTR_line);
+        top = prof_callback_invoke(callback, args);
+    }
+
+    // RETURN
+    const byte *ip = code_state->ip;
+    if (*ip == MP_BC_RETURN_VALUE || *ip == MP_BC_YIELD_VALUE) {
+        args->event = MP_ROM_QSTR(MP_QSTR_return);
+        top = prof_callback_invoke(callback, args);
+        if (code_state->prev_state && *ip == MP_BC_RETURN_VALUE) {
+            code_state->frame->callback = NULL;
+        }
+    }
+
+    // OPCODE
+    if (false) {
```

Comment:
A way to implement this would be to support setting `frame.f_trace_opcodes=True`, a feature added in CPython 3.7.